File: examples/c++/button_add.cpp
#include <string>
#include <map>

class Form {
public:
    std::map<std::string, std::string> fields;
    void clear() { fields.clear(); }
    std::string getField(const std::string& name) const {
        auto it = fields.find(name);
        return it != fields.end() ? it->second : "";
    }
    void setField(const std::string& name, const std::string& value) {
        fields[name] = value;
    }
};

class ButtonAdd {
    DataTable& table_;
    Form& form_;
public:
    ButtonAdd(DataTable& table, Form& form) : table_(table), form_(form) {}
    void onClick() {
        if (form_.getField("name").empty()) return;
        table_.addRecord(form_.fields);
        form_.clear();
    }
};

File: examples/python/button_add.py
import html as _html


class Form:
    def __init__(self):
        self.fields = {}

    def clear(self):
        self.fields.clear()

    def get_field(self, name):
        return self.fields.get(name, "")

    def set_field(self, name, value):
        self.fields[name] = value


class ButtonAdd:
    def __init__(self, table, form):
        self.table = table
        self.form = form

    def on_click(self):
        if not self.form.get_field("name"):
            return
        self.table.add_record(self.form.fields)
        self.form.clear()

    def render(self, columns):
        fields_html = "".join(
            f'<label>{_html.escape(c)}<input type="text" id="f_{_html.escape(c)}" required></label>'
            for c in columns
        )
        return f'''<button class="btn-add" onclick="showAdd()">+ Add Record</button>
<div class="modal-bg" id="addModal" onclick="if(event.target===this)this.style.display='none'">
<div class="modal">
<h3>Add Record</h3>
<form onsubmit="return saveAdd(event)">
{fields_html}
<div class="modal-btns">
<button type="button" class="cancel" onclick="document.getElementById('addModal').style.display='none'">Cancel</button>
<button type="submit" class="save">Save</button>
</div></form></div></div>'''
